### ISSUE TITLE
Connection fails, tested on v1.0.0-alpha6

### ISSUE THREAD

**Issue by Dango-rs** [sozu-proxy/lapin#271]
Decided to give the new tag a try and ran straight forward into a issue. Am I missing something with the new version?
Both tests were performed with https://github.com/sozu-proxy/lapin/blob/master/examples/consumer.rs
Tested with same settings on v0.39.7 (74c88c31cae02ab969e1879481cd47bb4336114c) and works like a charm, but on 1.0.0-alpha6 (5e5dfa9bf0dca58ad09552e231f156a24e4c3e98) it fails to connect to the server, returning the following:

> [2020-04-23T15:11:36Z ERROR lapin::io_loop] error writing: IOError(Os { code: 10057, kind: NotConnected, message: "A request to send or receive data was disallowed because the socket is not connected and (when sending on a datagram socket using a sendto call) no address was supplied." })
> [2020-04-23T15:11:36Z ERROR lapin::channels] Connection error
thread 'main' panicked at 'connection error: IOError(Os { code: 10057, kind: NotConnected, message: "A request to send or receive data was disallowed because the socket is not connected and (when sending on a datagram socket using a sendto call) no address was supplied." })', src\bin\main.rs:15:20

Backtrace:
```
stack backtrace:
   0: backtrace::backtrace::trace_unsynchronized
             at C:\Users\VssAdministrator\.cargo\registry\src\github.com-1ecc6299db9ec823\backtrace-0.3.40\src\backtrace\mod.rs:66
   1: std::sys_common::backtrace::_print_fmt
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\sys_common\backtrace.rs:77
   2: std::sys_common::backtrace::_print::{{impl}}::fmt
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\sys_common\backtrace.rs:59
   3: core::fmt::write
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libcore\fmt\mod.rs:1052
   4: std::io::Write::write_fmt<std::sys::windows::stdio::Stderr>
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\io\mod.rs:1426
   5: std::sys_common::backtrace::_print
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\sys_common\backtrace.rs:62
   6: std::sys_common::backtrace::print
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\sys_common\backtrace.rs:49
   7: std::panicking::default_hook::{{closure}}
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\panicking.rs:204
   8: std::panicking::default_hook
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\panicking.rs:224
   9: std::panicking::rust_panic_with_hook
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\panicking.rs:472
  10: std::panicking::begin_panic_handler
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\panicking.rs:380
  11: core::panicking::panic_fmt
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libcore\panicking.rs:85
  12: core::option::expect_none_failed
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libcore\option.rs:1199
  13: core::result::Result<lapin::close_on_drop::CloseOnDrop<lapin::connection::Connection>, lapin::error::Error>::expect<lapin::close_on_drop::CloseOnDrop<lapin::connection::Connection>,lapin::error::Error>
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\src\libcore\result.rs:991
  14: main::main::{{closure}}
             at .\src\bin\main.rs:16
  15: std::future::{{impl}}::poll<generator-1>
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\src\libstd\future.rs:43
  16: futures_executor::local_pool::{{impl}}::run_until::{{closure}}<std::future::GenFuture<generator-1>>
             at C:\Users\Someone\.cargo\registry\src\github.com-1ecc6299db9ec823\futures-executor-0.3.4\src\local_pool.rs:177
  17: futures_executor::local_pool::run_executor::{{closure}}<(),closure-0>
             at C:\Users\Someone\.cargo\registry\src\github.com-1ecc6299db9ec823\futures-executor-0.3.4\src\local_pool.rs:87
  18: std::thread::local::LocalKey<alloc::sync::Arc<futures_executor::local_pool::ThreadNotify>>::try_with<alloc::sync::Arc<futures_executor::local_pool::ThreadNotify>,closure-0,()>
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\src\libstd\thread\local.rs:262
  19: std::thread::local::LocalKey<alloc::sync::Arc<futures_executor::local_pool::ThreadNotify>>::with<alloc::sync::Arc<futures_executor::local_pool::ThreadNotify>,closure-0,()>
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\src\libstd\thread\local.rs:239
  20: futures_executor::local_pool::run_executor<(),closure-0>
             at C:\Users\Someone\.cargo\registry\src\github.com-1ecc6299db9ec823\futures-executor-0.3.4\src\local_pool.rs:83
  21: futures_executor::local_pool::LocalPool::run_until<std::future::GenFuture<generator-1>>
             at C:\Users\Someone\.cargo\registry\src\github.com-1ecc6299db9ec823\futures-executor-0.3.4\src\local_pool.rs:174
  22: main::main
             at .\src\bin\main.rs:15
  23: std::rt::lang_start::{{closure}}<()>
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\src\libstd\rt.rs:67
  24: std::rt::lang_start_internal::{{closure}}
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\rt.rs:52
  25: std::panicking::try::do_call<closure-0,i32>
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\panicking.rs:305
  26: panic_unwind::__rust_maybe_catch_panic
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libpanic_unwind\lib.rs:86
  27: std::panicking::try
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\panicking.rs:281
  28: std::panic::catch_unwind
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\panic.rs:394
  29: std::rt::lang_start_internal
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\/src\libstd\rt.rs:51
  30: std::rt::lang_start<()>
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447\src\libstd\rt.rs:67
  31: main
  32: invoke_main
             at d:\agent\_work\5\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
  33: __scrt_common_main_seh
             at d:\agent\_work\5\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
  34: BaseThreadInitThunk
  35: RtlUserThreadStart
```

**Comment by Keruspe** [comment]
I see you're running windows, this might be related I guess?

Would you be willing to bisect this issue?

```
git bisect start
git bisect good 0.39.7
git bisect bad 1.0.0-alpha6
```

Then run your test and use either `git bisect good` if your test passes or `git bisect bad` if it doesn't. This will take a few steps and will identify the commit introducing the regression for you.

I do not have any windows system to test this on my side, and on linux everything seems fine

**Comment by Keruspe** [comment]
Actually, as a first step you could try with 1.0.0-alpha5, as we switched to our new reactor in alpha6

**Comment by Dango-rs** [comment]
Alright so it seems that cargo has bugged the registry when I was previously switching from alpha1 to alpha6 because I have already tried them before writing here and they were failing as well. (no wonder the registry was so fast when updating..)

However 1.0.0-alpha5 doesn't seem to have that issue while 1.0.0-alpha6 have it.
Bisecting: 7 revisions left to test after this (roughly 3 steps)
[a1688fa6e94993d92cbd668b2060cbc45b827594] port to Reactor

e01108b4d9b3947eb3b1c997199342dca277004e is fine

Seems like io_loop::handle_write_result() is failing but socket status is can_read=true, can_write=true, has_data=true

**Comment by Keruspe** [comment]
Thanks for confirming. I'll see if I can find anything obvious, but might be kind of hand without being able to test myself

**Comment by Dango-rs** [comment]
I'll continue looking into the issue as well and see what I can dig up
tomorrow.

**Comment by Keruspe** [comment]
@Dango-rs Can you try the "connected" branch I just pushed?

**Comment by Dango-rs** [comment]
@Keruspe Connects successfully but drops the connection right after that.

Here's the full trace from the log. Basically the error didn't change, it just moved a step.

```
[2020-04-23T20:38:27Z TRACE amq_protocol_tcp] Connecting to xxxxxxxxxx.com:5672
[2020-04-23T20:38:27Z DEBUG lapin::channels] create channel with id 0
[2020-04-23T20:38:27Z TRACE lapin::channel] channel 0 send_frame
[2020-04-23T20:38:27Z TRACE lapin::channel] channel 0 wake
[2020-04-23T20:38:27Z TRACE mio::poll] registering event source with poller: token=Token(1), interests=READABLE | WRITABLE
[2020-04-23T20:38:27Z TRACE lapin::reactor] reactor poll
[2020-04-23T20:38:27Z TRACE pinky_swear] [ProtocolHeader.Ok] Called from future, registering waker
[2020-04-23T20:38:27Z TRACE lapin::io_loop] io_loop run
[2020-04-23T20:38:27Z TRACE pinky_swear] [ProtocolHeader.Ok] Called from future, registering waker up in chain
[2020-04-23T20:38:27Z TRACE lapin::socket_state] Got event for socket: Wake
[2020-04-23T20:38:27Z TRACE pinky_swear] [ProtocolHeader] Called from future, registering waker
[2020-04-23T20:38:27Z TRACE lapin::socket_state] Got event for socket: Wake
[2020-04-23T20:38:27Z TRACE lapin::io_loop] io_loop do_run; can_read=true, can_write=true, has_data=true
[2020-04-23T20:38:27Z INFO  lapin::io_loop] Connected to xxx.xxx.xxx.xxx:5672
[2020-04-23T20:38:27Z TRACE lapin::io_loop] will write to buffer: AMQPFrame::ProtocolHeader
[2020-04-23T20:38:27Z ERROR lapin::io_loop] error writing: IOError(Os { code: 10057, kind: NotConnected, message: "A request to send or receive data was disallowed because the socket is not connected and (when sending on a datagram socket using a sendto call) no ad
dress was supplied." })
[2020-04-23T20:38:27Z TRACE pinky_swear] [ProtocolHeader.Ok] Resolving promise
[2020-04-23T20:38:27Z TRACE pinky_swear] Got data, waking our waker
[2020-04-23T20:38:27Z ERROR lapin::channels] Connection error
[2020-04-23T20:38:27Z TRACE pinky_swear] [ProtocolHeader.Ok] Called from future, registering waker
[2020-04-23T20:38:27Z TRACE pinky_swear] [ProtocolHeader] Resolving promise
[2020-04-23T20:38:27Z TRACE pinky_swear] [ProtocolHeader.Ok] Called from future, registering waker up in chain
[2020-04-23T20:38:27Z TRACE pinky_swear] Got data, waking our waker
[2020-04-23T20:38:27Z TRACE pinky_swear] [ProtocolHeader] Called from future, registering waker
[2020-04-23T20:38:27Z TRACE pinky_swear] Got data, notifying next in chain
[2020-04-23T20:38:27Z TRACE pinky_swear] Got data, waking our waker
[2020-04-23T20:38:27Z TRACE pinky_swear] [ProtocolHeader] Dropping promise
[2020-04-23T20:38:27Z TRACE pinky_swear] [ProtocolHeader.Ok] Dropping promise
thread 'main' panicked at 'connection error: IOError(Os { code: 10057, kind: NotConnected, message: "A request to send or receive data was disallowed because the socket is not connected and (when sending on a datagram socket using a sendto call) no address was supp
lied." })', examples\consumer.rs:13:20
```

**Comment by Keruspe** [comment]
Pushed a new more exhaustive version following up on the same approach

**Comment by Keruspe** [comment]
Heh, I didn’t actually mean to auto close this one. Can you confirm the problem is gone with beta 1?

**Comment by Dango-rs** [comment]
@Keruspe ah, I thought you tested it successfully on another env that's running windows and that's why you closed it and I thought there's no more need to answer.

Yep I tested the same day you merged the extended commit and it was all good.
I just tested beta1 and its working fine both on my app and the example I quoted above.

Just one tiny thing that I saw:
```
[2020-04-25T18:28:07Z INFO  lapin::io_loop] Connected to xxx.xxx.xxx.xxx:5672
[2020-04-25T18:28:28Z ERROR lapin::channels] Connection error
thread 'main' panicked at 'connection error: IOError(Kind(ConnectionAborted))', examples\consumer.rs:13:20
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: process didn't exit successfully: `target\debug\examples\consumer.exe` (exit code: 101)
```

The success message is triggered too early, in fact its triggered when the connection hasn't been  established yet. Since crate previously wasn't declaring any logs on the info! macro and it was all up to the user to decide which information to log as info, I would suggest to remove the message. 

Anyways, I'd like to thank you for all the great work, as I forgot to do it previously in the initial post. Keep it up!
